### PR TITLE
Add Hooks APIs

### PR DESCRIPTION
**Pull request by blainekasten**
# @reach/router, we have the hooks

This PR Adds react-router compatible hook APIs:

- `useMatch(path: string): MatchBag`
- `useLocation(): Location`
- `useNavigate(): NavigateFn`
- `useParams(): { [pathPart: string]: string | number } | null`

Questions & Thoughts:

1. Should we consider naming something like `useRouterMatch` to avoid naming conflicts with user code for being overly generic?
1. ~Should `useMatch` return the `<Match>{bag =>}` object? I implemented it as a boolean which seems more usable and logical, but I might not know all the use cases.~

### CODE REVIEW

**Comment by ryanflorence** [comment]
Regarding (2), right, so:

```
// replaces getting stuff from component props
const bag = useRouteMatch()

// replaces (or is simply the implementation of) <Match/>
const match = useMatch(path)
```

**Comment by blainekasten** [comment]
@ryanflorence changed, and added documentation

**Comment by EloB** [comment]
What is the ETA on this one? I also want something like this. I mean is it worth waiting or make some quick and dirty until it's finished?

**Comment by blainekasten** [comment]
@EloB I would say by end of week is realistic, possibly sooner. Just hard to state clearly, we are trying to make sure the hooks align signature with future versions of react-router

**Comment by mjackson** [comment]
Hey @blainekasten, I just took a look at the hooks. Besides a few subtle differences between the names of things (e.g. `useHistory` vs `useNavigate`) it looks like you've got all the same functionality here as we have in React Router's hooks. Should be a very straightforward migration to RR v6 from here 👍

**Comment by mjackson** [comment]
BTW @blainekasten, I would probably rename `useMatch` to `useRouteMatch` like we're doing in React Router 5.1. I purposefully avoided using the name `useMatch` in the router because I was saving it for v6 where I expect that concept to change slightly. Happy to tell you more about it if you're curious. If not, just call it `useRouteMatch` and it'll serve more or less the same purpose as our current `useRouteMatch` and we will introduce a new `useMatch` hook in React Router v6 that will be slightly different.

File: website/src/markdown/api/useLocation.md [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+const AnalyticTracker = (props) => {
+  const location = useLocation();
+
+  useEffect(() => {
+    ga.send(['pageview', location.pathname]);
+  }, [])
+
+  return null
+)
+```
```
Comment (ryanflorence) [inline website/src/markdown/api/useLocation.md]: I'd just use a hook, we don't need these weird components that don't render anything anymore :D

```js
import { useLocation } from "@reach/router"

const usePageViews = () => {
  const location = useLocation();
  useEffect(() => {
    ga.send(['pageview', location.pathname]);
  }, [location]) // don't forget the location here:
)
```

**Comment by ryanflorence** [comment]
This says WIP but it looks good to me. Michael and I chatted about `useMatch` and we're good with it here, so I'm ready to merge this unless you've got some other stuff you needed to do?

File: src/index.js [hunk 2]
```diff
@@ -520,6 +520,56 @@ let Match = ({ path, children }) => (
...
+    throw new Error(
+      "useLocation hook was used but a LocationContext.Provider was not found in the parent tree. Make sure this is used in a component that is a child of Router"
+    );
+  }
+
+  return context.location;
+};
+
+const useNavigate = () => {
+  const context = useContext(LocationContext);
```
Comment (SimenB) [inline src/index.js]: Shouldn't all hooks have the `if (!context) {` guard?

Possibly by just having an internal `useLocationContext` and `useBaseContext` that are not exported to keep it a bit more DRY

Comment (blainekasten) [reply src/index.js]: Yeah why not. the above hook was pseudo-copied from an above render prop component. It's unlikely useNavigate and other hooks will be used in similar issues, but it doesn't hurt to add helpful warnings when tools are misconfigured. Happy to take a PR to fix this!

**Comment by blainekasten** [comment]
> This says WIP but it looks good to me. Michael and I chatted about useMatch and we're good with it here, so I'm ready to merge this unless you've got some other stuff you needed to do?

@ryanflorence damn it. I didn't see your message and changed it to `useRouterMatch` I'll switch it back

File: src/index.js (hunk 2 above)
Comment (blainekasten) [reply src/index.js]: Never mind, I took care of it! Thanks for the call out

**Pull request merged**